core-bundle/src/Security/Authentication/RememberMe/ExpiringTokenBasedRememberMeServices.php

```diff
@@ -0,0 +1,253 @@
+
+        $this->connection = $connection;
+        $this->secret = $secret;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function cancelCookie(Request $request): void
+    {
+        // Delete cookie on the client
+        parent::cancelCookie($request);
+
+        // Delete cookie from the tokenProvider
+        if (null !== ($cookie = $request->cookies->get($this->options['name']))
+            && 2 === \count($parts = $this->decodeCookie($cookie))
+        ) {
+            [$series] = $parts;
+            $this->connection->delete('tl_remember_me', ['series' => $series]);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processAutoLoginCookie(array $cookieParts, Request $request): UserInterface
+    {
+        if (2 !== \count($cookieParts)) {
+            throw new AuthenticationException('The cookie is invalid.');
+        }
+
+        [$series, $tokenValue] = $cookieParts;
+
+        try {
+            $this->connection->exec('LOCK TABLES tl_remember_me WRITE');
+
+            $rows = $this->loadTokens($series);
+
+            $currentToken = null === $rows[0]->expires ? $rows[0] : null;
+            $matchedToken = $this->findValidToken($rows, $tokenValue);
+
+            $this->cleanupExpiredTokens();
+
+            if (null === $currentToken || $currentToken->value === $matchedToken->value || null !== $matchedToken->expires) {
+                $tokenValue = $this->migrateToken($matchedToken);
+            } else {
+                $tokenValue = $currentToken->value;
+            }
+        } catch (\Exception $e) {
+            if ($e instanceof AuthenticationException) {
+                throw $e;
+            }
+
+            throw new AuthenticationException('Rememberme services failed.', 0, $e);
+        } finally {
+            $this->connection->exec('UNLOCK TABLES');
+        }
+
+        $request->attributes->set(self::COOKIE_ATTR_NAME, $this->createCookie($request, $series, $tokenValue));
+
+        return $this->getUserProvider($matchedToken->class)->loadUserByUsername($matchedToken->username);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function onLoginSuccess(Request $request, Response $response, TokenInterface $token): void
+    {
+        $series = base64_encode(random_bytes(64));
+        $tokenValue = base64_encode(random_bytes(64));
+
+        $this->insertToken(
+            \get_class($user = $token->getUser()),
+            $user->getUsername(),
+            $series,
+            $tokenValue
+        );
+
+        $response->headers->setCookie($this->createCookie($request, $series, $tokenValue));
+    }
+
+    private function loadTokens(string $series): array
+    {
+        try {
+            $rows = $this->connection->executeQuery(
+                'SELECT * FROM tl_remember_me WHERE series=:series AND (expires IS NULL OR expires>=:expires) ORDER BY expires IS NULL DESC',
```

Comment:
You mean `now` instead of `expires`?

Reply:
Yes. It's easer to read when you write queries like this :) If you read it out loud, right now it is `where series equals series AND (expires is null or expires is greater than or equal to expires)`. It doesn't really say much. This would read much better: `where series equals series AND (expires is null or expires is greater than or equal to now)`.

Reply:
There is also a built-in `NOW()` function in MySQL.